hpx/parallel/util/detail/chunk_size.hpp

```diff
@@ -87,53 +131,95 @@ namespace hpx { namespace parallel { namespace util { namespace detail
+        
+        // make sure we have enough work left to actually run the benchmark
+        if(count < test_chunk_size * cores) return 0;
+        
+        // run the benchmark iterations
+        std::vector<Future> bench_futures(cores);
         boost::uint64_t t = hpx::util::high_resolution_clock::now();
-        add_ready_future_idx(workitems, f1, base_idx, first, test_chunk_size);
-
-        t = (hpx::util::high_resolution_clock::now() - t) / test_chunk_size;
+        for(std::size_t i = 0; i < cores; i++){
+            bench_futures[i] = std::move(hpx::async(f1, base_idx,
+                                                    first, test_chunk_size));
+            // mark benchmarked items as processed
+            base_idx += test_chunk_size;
+            std::advance(first, test_chunk_size);
+            count -= test_chunk_size;
+        }
+        hpx::wait_any(bench_futures); 
+        t = (hpx::util::high_resolution_clock::now() - t);
+        
+        for(std::size_t i = 0; i < cores; i++){
+            workitems.push_back(std::move(bench_futures[i]));
+        }
 
-        base_idx += test_chunk_size;
-        std::advance(first, test_chunk_size);
-        count -= test_chunk_size;
+        // get the timer step size
+        boost::uint64_t t_min = hpx::util::high_resolution_clock::min();
+        
+        // subtract 300 (constant measurement overhead)
+        if(t <= 300)
+        {
+            t = 0;
+        }
+        else
+        {
+            t -= 300;
+        }
+        
+        // if time was smaller than being able to measure, consider it to be
+        // the smallest possible amount of time. this will get important,
+        // and is an approximation necessary to prevent the creation of too
+        // many asyncs.
+        if(t <= t_min) t = t_min;
+        
+        // calculate number of chunks, round
+        std::size_t num_chunks_dividend = static_cast<std::size_t>
+                            (t * count);
+        std::size_t num_chunks_divisor  = static_cast<std::size_t>
+                            (test_chunk_size * desired_chunktime_ns.count());
+        std::size_t num_chunks =
+                            (num_chunks_dividend + num_chunks_divisor / 2) 
+                                                        / num_chunks_divisor;
+        
+        // if benchmark returned smallest amount of time, prevent creation
+        // of too many asyncs and do normal geometric distribution
+        if((t <= t_min) && num_chunks > cores)
+            num_chunks = cores;
+        
+        // prevent 0 chunks
+        if(num_chunks == 0) num_chunks = 1;
+        
+        // calculate desired chunksize from number of chunks, ceil
+        std::size_t chunksize = (count + num_chunks - 1) / num_chunks; 
+        
+        return chunksize;
 
-        // return chunk size which will create 80 microseconds of work
-        return t == 0 ? 0 : (std::min)(count, (std::size_t)(80000 / t));
     }
 
```

Comment:
> Hi, I was looking over the thing and doing it like this sample might work: 
https://gist.github.com/koonschi/adb578aab970bf41f3c1

That's not what I meant. I would like to replace the two different (nearly identical) functions with only one (variadic) function. That should be possible in this case as the two existing functions differ mainly in an additional argument which is passed through to a user provided function.